md/documentation-template.j2

```diff
@@ -0,0 +1,1273 @@
+- Callable (Executor Service)
+- IFunction (Atomics)
+- EntryProcessor (JCache)
+- ExpiryPolicy (JCache)
+- CacheConfig (JCache)
+- ListenerConfig (JCache)
+- Interceptor (Map)
+
+A client may follow Hazelcast's native serialization or it may implement its own custom serialization solution. For more
+information on how Hazelcast serializes its objects, see the official [Reference Manual](https://docs.hazelcast.org/docs/latest/manual/html-single/#serialization) serialization section.
+
+For all byte-array parameters, the API users should implement the following, depending on the operation type:
+- If the operation is such that no member side deserialization is needed for the parameter, then the user can just use
+any serialization and there is no need for implementation on the member side.
+- If the operation processing at the member requires deserialization of the byte-array parameter, then the user should
+use a Java object implementing one of the Hazelcast serializations and may need to do some member side implementations
+depending on the chosen serialization type. Furthermore, the serializer must be registered in the serialization configuration
+of the member as described in the serialization section of the [Reference Manual](https://docs.hazelcast.org/docs/latest/manual/html-single/#serialization).
+
+The client authentication request message contains the serialization version parameter. The client and the member decide the serialization
+version to be used using this information. Hazelcast serialization versions will be matched to provide a compatible serialization.
+There are two cases that can occur:
+- The client may have a higher serialization version than the member. In that case, the client will auto configure
+itself during authentication to match the member serialization version.
+- The client may have a lower serialization version than the member. In that case, the member should be configured with
+the system property to downgrade the member serialization version.
+
+### 6.3. Security
+Most of the security is configured on the member side in a Hazelcast cluster. A client must authenticate itself, which in turn
+lets the member establish an endpoint for the client. The member can restrict what a client can and cannot access. The current protocol
+does not provide explicit support for encryption. For more information, see the [Security](https://docs.hazelcast.org/docs/latest/manual/html-single/#security) chapter of the Hazelcast Reference Manual.
+
+{# END of manually written sections #}
+## {{ protocol_message_offset }}. Protocol Messages
+### {{ protocol_message_offset }}.1. Custom Data Types Used In The Protocol
+{% for definition in custom_definitions %}
+{% set custom_types = definition.get('customTypes', None) %}
+{% if custom_types is not none %}
+{% for custom_type in custom_types %}
+#### {{ protocol_message_offset }}.1.{{ loop.index }}. {{ custom_type.name }}
+**Parameters**
+
+| Name | Type | Nullable | Available Since |
```

Comment:
nit: it might be a good idea to add descriptions for parameters in custom types. This could be tracked in a separate GH issue.

Reply:
Created an issue for it https://github.com/hazelcast/hazelcast-client-protocol/issues/369